bugfix/618

tedrek opened the pull request:
Resolves #618 

Fixes aircraft in cruise phase not descending if they have no waypoints with an AT altitude restriction while on a STAR.  Now aircraft should start their descent correctly just so long as some altitude restriction is in place.

n8rzz commented on src/assets/scripts/client/aircraft/AircraftModel.js:
```diff
@@ -848,15 +850,26 @@ export default class AircraftModel {
+              waypointsWithRelevantCeiling[0].altitudeMaximum;
+        const targetPosition = _isEmpty(waypointsWithRelevantCeiling) ?
+              _last(this.fms.waypoints).positionModel :
+              waypointsWithRelevantCeiling[0].positionModel;
```
If you're using `_last` to get the last item in an array you should also use `_head` to get the first item of an array.  Consistency.

https://lodash.com/docs/4.17.4#head

n8rzz commented on src/assets/scripts/client/aircraft/AircraftModel.js:
```diff
@@ -1785,7 +1798,12 @@ export default class AircraftModel {
+                if (_isNil(hardRestrictedWaypointModel)) {
+                    return this.fms.getBottomAltitude();
+                }
+                else {
```
this `else` in unnecessary.  if we reach the `if` block above it we'll be `returned` so anything after that is considered the `else`

_another fair warning: I'm **really** not a fan of `else`.  logic should be clean with a single `if` and `return` instead of an `if`/`else` whenever possible_

n8rzz commented on src/assets/scripts/client/aircraft/AircraftModel.js:
```diff
@@ -1827,22 +1845,23 @@ export default class AircraftModel {
+        if (nextRestrictedWaypoint.altitudeMinimum !== INVALID_NUMBER) {
+            return nextRestrictedWaypoint.altitudeMinimum;
+        }
+        else {
```
`else` is not needed here.  

Also, `else` should be on the same line as the closing bracket of the previous logical statement:

```javascript
// undesired
if (condition)
{
  // do stuff
}
else
{
  // do other stuff
}

if (condition) {
  // do stuff
}
else {
  // do other stuff
}

// desired
if (condition) {
  // do stuff
} else {
  // do other stuff
}

if (condition) {
  // do stuff
  return
}
```

Be sure you are using `eslint` with your editor: https://eslint.org/docs/user-guide/integrations.  It will warn you about these style inconsistencies.

n8rzz commented on test/aircraft/AircraftModel.spec.js:
```diff
@@ -60,3 +61,23 @@ ava('.getViewModel() includes an altitude that has not been rounded to the neare
     t.true(result === 77.77123456700001);
 });
+
+ava('.updateTarget() causes arrivals to descend when the STAR includes only AT or ABOVE altitude restrictions', (t) => {
```
_Soooo_ nice to see some more tests for the `AircraftModel` 👍

tedrek commented:
@n8rzz, I've addressed most of your comments with the new commits.  Thanks for the feedback, I'm a sysadmin by trade who started with perl so... fair warning :wink:.  I will work on getting ESLint setup on my machines and try to conform to the rest of the style bits.
For refactoring the logic out to Fms, my feeling is the correct way to do that would be for the Fms to calculate all of this once and create it's own waypoints for descent points, speed changes, etc.  to comply with the waypoints in the procedure.  I don't have the time at the moment to tackle such a large change, if you want I can still move the current logic over.
I also added a hasMinimumAltitudeAbove() method to WaypointModel, it's not technically required but the existence of the necessary hasMaximumAltitudeBelow() method implies the former and it's really easy to write it now (and test it).  Would appreciate your thoughts there.

n8rzz commented:
Mmm, Perl.  Those were the days.  We actually have a project here that is still using some Perl, if you can believe it.

I agree with you there, it would be much more efficient to calculate that once.  I worry about then needing to maintain another list of waypoints.  Maybe that would be rolled into the existing waypoint model?  I'd be interested to see how you solve that.

We have two possible paths to take here:
1) you can take the time you need to do all this in one shot as part of this PR
1) you can leave the future Fms bits in this PR and open a new issue for moving it to the Fms, then add `TODO` comments referencing that issue and that the logic is planned to move.

These points depend on how anxious Mr @erikquinn is to get this functionality into the sim 😉 

I like the forward thinking with the additional methods 👍.  A guideline we like to follow is that any new methods that contain branching logic should also have tests for each logical branch.  I believe you have some tests for those already, but I wanted to call that out here just so you are aware.  As I'm sure you can imagine, with an app this complex, tests are your friend.  Its much easier to write the tests as you write the code than it is to go back later.

So yeah, if you feel we'll need it, add it!  Be sure to test it, comment it and add notes if you feel the method needs them.

erikquinn commented:
Definitely door 2 please, if you don't mind. This one is a fairly critical bugfix that needs to make it into this sprint (aka merged by Monday the 28th). Without it, MDSD is broken and a couple other airport PRs are blocked. And I have some thoughts on how a whole bunch of things should actually be reworked in this area, so it's better to push the big refactors to next month.

tedrek commented:
I think this everything is addressed at this point as consensus seems to be option 2, @n8rzz could you give it a check over again?

@erikquinn, I've add #656 for the Fms refactor bits, I would be interested in your thoughts on any refactoring in that area as it's something I have some interest in as well.

erikquinn commented on src/assets/scripts/client/aircraft/FlightManagementSystem/WaypointModel.js:
```diff
@@ -368,6 +368,32 @@ export default class WaypointModel {
+     * @param altitude {number} in feet
+     * @return {boolean}
+     */
+    hasMaximumAltitudeBelow(altitude) {
```
At first glance, I feel like these should be just `.hasMaximumAltitude()`. It would seem strange to have `.hasMaximumAltitudeBelow(altitude)` without a way to see if there simply is one or not. However, we do get the benefit here of excluding any restrictions that are above the aircraft's cruise altitude, which is something I wouldn't have thought of and would definitely want to keep in some form. Maybe this _is_ best to keep it this way? :thinking:

tedrek replied:
I suspect there is a more elegant solution along the lines of a `.hasMaximumAltitude()` and `.hasMaximumAltitudeFiltered()` but I can't come up with the right words to convey the semantics of such a thing.  For now it's probably a good compromise.

erikquinn replied:
Works for now! :smiley:

n8rzz commented on src/assets/scripts/client/aircraft/AircraftModel.js:
```diff
@@ -840,6 +844,7 @@ export default class AircraftModel {
         return decelerationTime > timeUntilWaypoint;
     }
 
+    /* TODO: Refactor all this logic into the Fms - #656
```
here it actually does matter.  one line comments in  js should be with `//` and multi-line with: ```
/* 
 *
 */
``` A multi-line comment also needs to be closed with the `*/` at the end of it.

tedrek replied:
That's just a stupid mistake on my part... which eslint apparently didn't pick up.  Fixed

n8rzz commented on src/assets/scripts/client/aircraft/AircraftModel.js:
```diff
@@ -848,15 +853,23 @@ export default class AircraftModel {
-        const altitudeChange = waypointAltitude - this.altitude;
+        const currentAltitude = this.altitude;
+        const waypoints = this.fms.getAltitudeRestrictedWaypoints();
+        const waypointsWithRelevantCeiling = _filter(waypoints, (waypoint) => waypoint.hasMaximumAltitudeBelow(this.altitude));
```
Wow, you're right, we don't have a `LegCollection`.  That surprised me very much.

tedrek replied:
My opinion is the really proper way of doing it is to encapsulate effectively all the logic within the Fms because minimum altitudes are also a factor.  I'm unclear if you would like a change within this pull request or if it's more for the future refactor.

erikquinn replied:
I agree it should be moved out of the aircraft model. But no, I only meant these as thoughts for the future. This is good for now.

tedrek commented on src/assets/scripts/client/aircraft/AircraftModel.js:
```diff
@@ -848,15 +853,23 @@ export default class AircraftModel {
+              _head(waypointsWithRelevantCeiling).altitudeMaximum;
+        const targetPosition = _isEmpty(waypointsWithRelevantCeiling) ?
+              _last(this.fms.waypoints).positionModel :
+              _head(waypointsWithRelevantCeiling).positionModel;
```
I will keep the verbosity with variables in mind, it's something I should definitely do more.

The example quoted here is not equivalent to the pull request code.  The intent is to check for the next waypoint with a Maximum altitude restriction which the aircraft needs to descend to in order to comply with, in the case there are no maximum altitude restrictions then the descent should be to the lowest minimum altitude restriction and should occur as late as possible to reach that altitude while still on the procedure.  This doesn't properly take into account the somewhat unlikely case where a procedure has a minimum restriction which must be met followed by a maximum restriction and the aircraft is unable to descend fast enough between the two to comply with both.  There is actually a couple of possible variations on that edge case which aren't quite properly dealt with.

As for the logic, this is one of the rare cases where a ternary may be the best option locally.  A guard statement isn't appropriate because we actually want to take one of two branches.  A bunch of the relevant state is in the current function so it's not an easy refactor into Fms.  The alternative to the ternary locally is an `if` statement but then we can't use the `const` keyword as it's block scoped so it becomes a `let` outside of the if which is overwritten only once on some procedures.

erikquinn replied:
Indeed the quoted example is not equivalent to the pull request code, because there are some issues with how that logic works out:

> in the case there are no maximum altitude restrictions then the descent should be to the lowest minimum altitude restriction 

this _is_ true, but also not really relevant to this particular method. If there are no further relevantly-restricted waypoints, then it doesn't matter what the bottom altitude is... we're past the point where there's any useful descent planning information, so the answer to the question of whether or not we are `past the top of descent` really can't be answered. And so, we can just return. Also, `_last(this.fms.waypoints)` is probably not the one you're looking for to match the bottom altitude anyway.

In general, for our purposes, overwriting a `let` declared variable would be highly preferable to a ternary, especially in cases like this where multiple ternaries are used contingent to a common condition.

> This doesn't properly take into account the somewhat unlikely case where a procedure has a minimum restriction which must be met followed by a maximum restriction and the aircraft is unable to descend fast enough between the two to comply with both. 

And this is true as well. That would be nice to have the aircraft advise the controller that they are unable to make the second restriction, but that's a shortcoming that isn't a big concern since most procedures will be designed to be conducive to performance of the aircraft frequently using it. Either way though, this is something that would be dealt with in the `_calcualateTargetAltitude....()` areas or maybe future methods where the descent planning is done for all waypoints, but is out of scope of this particular method, which should only be concerned with whether the FMS should remain in cruise mode or kick into descent mode.

tedrek replied:
The ternary has been removed and replaced with `let` and re-assignment.

I've renamed the method to `shouldStartDescent` as that's really what it's being used to decide.  As such, it should start the descent in two scenarios:
1. There is a maximumAltitude restriction and the aircraft needs to descend to meet it.
2. A continuous descent along the planned flight path is required to intercept the expected approach.

The current implementation of the flight plan doesn't actually terminate at an airport or runway so I'm doing the next best thing and starting the descent at the point where a continuous descent would reach the minimum altitude at the end of the flight plan in case of scenario 2.  The Fms also doesn't know where the aircraft is so can't tell the distance remaining to any given waypoint, which means doing the direct distance to the last waypoint.

tedrek commented on src/assets/scripts/client/aircraft/AircraftModel.js:
```diff
@@ -1769,13 +1782,13 @@ export default class AircraftModel {
             case FLIGHT_PHASE.CLIMB:
+                if (_isNil(hardRestrictedWaypointModel)) {
+                    return;
+                }
```
The whole thing needs refactoring, but there isn't any bugs open against the climb behaviour as the logic on line 1187 causes it to climb to the intended ceiling anyway I believe.

erikquinn replied:
I'm away from a usable comptuer at the moment... I guess my concern is what would happen if we had a SID that contained only soft restrictions? Would it comply with those restrictions as it climbed up to the `initial_altitude`? Or would it not recognize there being any restrictions since none of them are _hard restrictions_, and just blast straight up for the initial_alt?

Setting one SID to having various nonsensical altitude restrictions and the aircraft to having a climb rate of like 50,000fpm should get us the answer. :wink:

tedrek replied:
I suspect it doesn't behave correctly, but it does probably behave playably at the moment ;)  I'll see about writing a couple of tests for the refactor.

tedrek commented on src/assets/scripts/client/aircraft/AircraftModel.js:
```diff
@@ -1785,6 +1798,10 @@ export default class AircraftModel {
                 // will enter the below block because the flight phase will have become 'DESCENT'.
+                if (_isNil(hardRestrictedWaypointModel)) {
+                    return this.fms.getBottomAltitude();
+                }
```
If there is no AT restrictions there may still be AT or ABOVE restrictions which can set a bottom altitude, similiarly there may be AT or BELOW restrictions which should be complied with.

tedrek replied:
Err this is also slightly hacky in the sense it intentionally just starts the aircraft descending and leaves the details to the descent phase logic.

erikquinn replied:
Right, and upon initial descent, the pilot would have set the MCP altitude (new) to the bottom altitude, so here we should be referencing not the bottom altitude from the fms, but rather `this.mcp.altitude`.

tedrek replied:
I've changed it to return the same thing it would after descent starts.

tedrek commented on src/assets/scripts/client/aircraft/AircraftModel.js:
```diff
@@ -1827,22 +1844,22 @@ export default class AircraftModel {
+        const nextRestrictedWaypoint = this.fms.nextAltitudeRestrictedWaypoint;
+        if (_isNil(nextRestrictedWaypoint)) {
+            return this.fms.getBottomAltitude();
+        }
```
Probably right here, not sure what the really correct thing is at the moment

erikquinn replied:
After thinking about it, yes, i think it is `this.mcp.altitude` we should be looking at here.

tedrek replied:
I believe `this.fms.getBottomAltitude()` is the correct thing here.  It won't return anything useful right now because the FMS doesn't actually plan the bottom altitude so it will do nothing currently.  However in the future the FMS should return the correct value.
`this.mcp.altitude` is likely to hold an incorrect value.  There are basically two times it would be used, and both occur with at least one waypoint at the end of the procedure without an altitude restriction.  The first is an automatic descent from cruise altitude which would currently leave the cruise altitude in the MCP.  The second is if a controller issued an altitude clearance and later issued a descend via star clearance,  the MCP would hold either an invalid number or the altitude clearance.

erikquinn commented:
@n8rzz Do you mind if I merge this as-is? It does seem to fix the problem, and this area will be revisited soon anway. I'd like to get MDSD fixed, and allow us to also add @Fechulo's KSTL.

n8rzz commented:
@erikquinn works for me :shipit:

erikquinn commented on src/assets/scripts/client/aircraft/AircraftModel.js:
```diff
@@ -840,23 +844,35 @@ export default class AircraftModel {
+        const currentAltitude = this.altitude;
+        const waypoints = this.fms.getAltitudeRestrictedWaypoints();
+        const waypointsWithRelevantCeiling = _filter(waypoints,
+                                                     (waypoint) => waypoint.hasMaximumAltitudeBelow(currentAltitude));
```
done

erikquinn commented on src/assets/scripts/client/aircraft/AircraftModel.js:
```diff
@@ -1823,26 +1839,25 @@ export default class AircraftModel {
-        return Math.max(waypointAltitude, this.mcp.altitude);
+
+        const nextRestrictedWaypoint = this.fms.nextAltitudeRestrictedWaypoint;
+        if (_isNil(nextRestrictedWaypoint)) {
```
done

erikquinn commented on src/assets/scripts/client/aircraft/AircraftModel.js:
```diff
@@ -1823,26 +1839,25 @@ export default class AircraftModel {
+
+        const nextRestrictedWaypoint = this.fms.nextAltitudeRestrictedWaypoint;
+        if (_isNil(nextRestrictedWaypoint)) {
+            return this.fms.getBottomAltitude();
```
done
